### PR TITLE
[BFD-1049] - patient endpoint v2 enhancments for AB2D and paging

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-1049](https://jira.cms.gov/browse/BFD-1049)

**User Story or Bug Summary:**

User Story:

As AB2D, I need for the patient endpoint in v2 to be as performant as it is in V1, so that my end users can migrate to v2 when we release it in July.  
A/C:

The following changes that were made to improve Patient endpoint performance in V1 are implemented in V2: 
https://github.com/CMSgov/beneficiary-fhir-data/pull/647
https://github.com/CMSgov/beneficiary-fhir-data/pull/668
https://github.com/CMSgov/beneficiary-fhir-data/pull/672
https://github.com/CMSgov/beneficiary-fhir-data/pull/673
In V2, the "next page" indicator displays only when there are additional results on paged responses

---

### What Does This PR Do?

User Story:

As AB2D, I need for the patient endpoint in v2 to be as performant as it is in V1, so that my end users can migrate to v2 when we release it in July.  

### What Should Reviewers Watch For?

Make sure the paging works as expected and has the changes that are associated with V1
If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [ X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X ] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X ] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

N/A

### Submitter Checklist

I have gone through and verified that...:

* [ X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X ] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [ X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 1]
```diff
@@ -1103,50 +1030,163 @@ public Bundle searchByCoverageContractAndYearMonth(
...
     }
   }
+
+  /**
+   * Check that the page size is valid
+   *
+   * @param paging to check
+   */
+  public static void checkPageSize(LinkBuilder paging) {
+    if (paging.getPageSize() == 0) throw new InvalidRequestException("A zero count is unsupported");
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:1189]: couldn't we just combine this exception? i.e., paging.getPageSize() < 1 : "page size less than 1 (one) is not supported"

**Review by mackec1 — APPROVED** [review]
LGTM....this will get further cleaned up with new schema/app changes.

**Review by mackec1 — DISMISSED** [review]
ship it!

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 2]
```diff
@@ -142,10 +150,14 @@ public Patient read(@IdParam IdType patientId, RequestDetails requestDetails) {
     CriteriaQuery<Beneficiary> criteria = builder.createQuery(Beneficiary.class);
     Root<Beneficiary> root = criteria.from(Beneficiary.class);
 
-    // commented out as in V2 code;  keep it that way for now
-    // if (requestHeader.isMBIinIncludeIdentifiers())
     root.fetch(Beneficiary_.medicareBeneficiaryIdHistories, JoinType.LEFT);
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:153]: yea, we don't support include MBI identifiers in v2.

[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:153]: Ok, then should we not have the line below (line 158-160)?

**Review by christrotteradhoc — APPROVED** [review]

**Pull request merged**